### ISSUE TITLE
Using mask parameter in skimage.segmentation.watershed

### ISSUE THREAD

**Issue by drchrisch** [scikit-image/scikit-image#7074]
### Description:

I was playing with the "mask" parameter from skimage.segmentation.watershed and got a somewhat unexpected result. Starting from the example given in the documentation I created a new mask with the intention to restrict object detection to mask==True regions. However, when the mask region is dilated the detected objects were larger than expected from the distance input values. 
Is this the intended behaviour, do I misunderstand the documentation or is this a bug?

```python
import matplotlib.pyplot as plt
from scipy import ndimage as ndi
from skimage.segmentation import watershed
from skimage.feature import peak_local_max
from skimage.morphology import disk

# Generate an initial image with two overlapping circles
x, y = np.indices((80, 80))
x1, y1, x2, y2 = 28, 28, 44, 52
r1, r2 = 16, 20
mask_circle1 = (x - x1)**2 + (y - y1)**2 < r1**2
mask_circle2 = (x - x2)**2 + (y - y2)**2 < r2**2
image = np.logical_or(mask_circle1, mask_circle2)

# Now we want to separate the two objects in image
# Generate the markers as local maxima of the distance to the background
distance = ndi.distance_transform_edt(image)
coords = peak_local_max(distance, footprint=np.ones((3, 3)), labels=image)
mask = np.zeros(distance.shape, dtype=bool)
mask[tuple(coords.T)] = True
markers, _ = ndi.label(mask)

labels = watershed(-distance, markers, mask=image, watershed_line=True)

### xtra code, prepare mask from marker positions by dilation with fairly large structuring element
disk_radius = 23
mask_xtra = ndi.binary_dilation(mask, structure=disk(disk_radius, dtype=np.uint8, strict_radius=True), iterations=1, border_value=0)
labels_xtra = watershed(-distance, markers, mask=mask_xtra, watershed_line=True)
### xtra code

fig, axes = plt.subplots(ncols=5, figsize=(13, 3), sharex=True, sharey=True)
ax = axes.ravel()

ax[0].imshow(image, cmap=plt.cm.gray)
ax[0].set_title('Overlapping objects')
ax[1].imshow(-distance, cmap=plt.cm.gray)
ax[1].set_title('Distances')
ax[2].imshow(labels, cmap=plt.cm.nipy_spectral)
ax[2].set_title('Separated objects')
ax[3].imshow(mask_xtra, cmap=plt.cm.gray)
ax[3].set_title('Mask 2')
ax[4].imshow(labels_xtra, cmap=plt.cm.nipy_spectral)
ax[4].set_title('Separated objects (mask xtra)')

for a in ax:
    a.set_axis_off()

fig.tight_layout()
plt.show()
```

![watershed](https://github.com/scikit-image/scikit-image/assets/42037938/7e14e943-d39e-40bd-8f9d-86d1d4159626)

### Way to reproduce:

```
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print(f'scikit-image version: {skimage.__version__}')
import numpy as np; print(f'numpy version: {np.__version__}')
import scipy; print(f'scipy version: {scipy.__version__}')
import matplotlib; print(f'matplotlib version: {matplotlib.__version__}')
```

### Version information:

```Shell
3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 17:59:51) [MSC v.1935 64 bit (AMD64)]
Windows-10-10.0.19045-SP0
scikit-image version: 0.21.0
numpy version: 1.25.1
scipy version: 1.11.1
matplotlib version: 3.7.2
```

**Comment by lagru** [comment]
Hey, thanks for the report and sorry for the late response from our side. I think the behavior that your example demonstrates is intended. The intensity topology of the image mainly impacts the order with which watersheds from different markers fill regions. 

E.g. inside the helper function `_validate_inputs` the parameter is described as https://github.com/scikit-image/scikit-image/blob/5408455469ddf3bc9456a6787f903b69a3a52316/skimage/segmentation/_watershed.py#L37-L38

and is set to `mask = np.ones(image.shape, bool)` if `mask` is not given. The Cython implementation also seems to confirm this. 

Though, I totally think that we could do better regarding the parameter description in the docstring. Would you like to create a PR updating the docstring so that it is more clear to you? I'd gladly review and help with a PR. :)

**Comment by drchrisch** [comment]
Hey, thanks for responding.
Going through code and examples, I realized that I misunderstood the flooding performed by the watershed algorithm. Saying that, I could imagine to modify the "mask" argument description: "Array of same shape as _image_. Defaults to ndarray of 1s. May be used to restrict labeling to thresholded regions (see **Examples** section)".

**Comment by lagru** [comment]
Sounds good. Go for it if you want and feel welcome to request a review from me! :D

**Comment by drchrisch** [comment]
Ok, how do I assign you as reviewer?

**Comment by mkcor** [comment]
I'll close this, since it was addressed by PR #7135, which itself superseded #7130 and was closed following the conversation which happened there.
